src/main/java/httpServer/ListenerAndResponder.java

```diff
@@ -40,46 +43,114 @@ public Socket keepListening() throws IOException {
-                if (line.equals("")) { break; }
+            while ((character = bufferedReader.read()) != null) {
+
+                headers.append(character);
+                if (checkIfReachedEndOfHeaders(headers)) {
+                    break;
+                }
             }
-            return stringBuilder.toString();
+
         } catch (IOException e) {
             e.printStackTrace();
-            return stringBuilder.toString();
+            return requestBuilder.build();
+        }
+
+        if (checkIfHeadersIndicateARequestBody(headers)) {
+            int bodyLength = getBodyLength(headers);
+            try {
+                while ((character = bufferedReader.read()) != null) {
+                    if (java.util.Objects.equals(bodyLength, 0)) {
+                        break;
+                    }
+                    body.append(character);
+                    if (body.length() == bodyLength) {
+                        break;
+                    }
+                }
+            } catch (IOException e) {
+                e.printStackTrace();
+                return requestBuilder.build();
+            }
+        }
+        return turnStringBuilderIntoHttpRequest(headers, body, requestBuilder);
+    }
+
+    private HTTPRequest turnStringBuilderIntoHttpRequest(StringBuilder headers, StringBuilder body, RequestBuilder requestBuilder) {
+        String[] headersArray = headers.toString().split(crlf);
+        for (String header: headersArray) {
+            requestBuilder = discernHeaderTypeAndAddToRequestBuilder(header, requestBuilder);
+        }
+        if (body.length() != 0) {
+            requestBuilder.setBody(body.toString());
         }
+        return requestBuilder.build();
     }
+
+    private RequestBuilder discernHeaderTypeAndAddToRequestBuilder(String line, RequestBuilder requestBuilder) {
+        if (!line.contains(": ")) {
+            requestBuilder.buildRequestLine(line);
+        } else {
+            requestBuilder.buildHeaderLine(line);
+
+        }
+        return requestBuilder;
+    }
+
+    private boolean checkIfReachedEndOfHeaders(StringBuilder requestScannedInSoFar) {
+        return requestScannedInSoFar.toString().contains(crlf + crlf);
+    }
+
+    private boolean checkIfHeadersIndicateARequestBody(StringBuilder headers) {
+        String headersAsString = headers.toString();
+        if (headersAsString.contains("Content-Length:")) {
+            if (getBodyLength(headers) == 0) {
+                return false;
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private int getBodyLength(StringBuilder headers) {
+        int indexOfContentLengthHeader = headers.indexOf("Content-Length: ");
+        int indexOfContentLength = indexOfContentLengthHeader + 16;
+        String trimmedHeaders = headers.substring(indexOfContentLength);
+        String lengthOfHeaders = trimmedHeaders.split(crlf)[0];
+        return Integer.parseInt(lengthOfHeaders);
```

Comment:
This feels unnecessarily complicated. Can you split on `crlf`, find the element in the resulting array that contains `Content-Length` and then split that string element on `: ` to get the length?

Reply:
Fixed it... I think. It still has a loop, which is a little cumbersome, but I wasn't sure how to do something like `.forEach` with a string array

Reply:
You can convert `String[]` to a `List` then use `.stream()` to convert to a `Stream` and then use `.filter()` (https://docs.oracle.com/javase/8/docs/api/java/util/stream/Stream.html#filter-java.util.function.Predicate-) to find the right element.

Reply:
I'm having trouble with the syntax of this one, will return to this